### PR TITLE
[5.0] Add new Verify Post Size middleware

### PR DESCRIPTION
**Pull request by laurencei**
So I was doing some AJAX file uploading, and kept getting random `TokenMismatchException` errors. After much debugging - I realised the issue related to `post_max_size`.

If the file I was attempted to upload was greater than my `post_max_size` - PHP clears the super globals of `$_POST` and `$_FILE` - **but** still sends the server the POST request (which is now empty).

The problem is that the empty post hits Laravel, and when doing CSRF detection there is no `_token` - so you'll get a `TokenMismatchException`, leading you on wild goose chases as to why your forms are not sending correct tokens.

It also means that the user will be shown an incorrect error. Rather than be told they are uploading a file that is _way_ to large - it just throws a 500 error with the wrong message.

I did some investigating - and there is a way to detect this situation. I propose we include this as part of the HTTP Middleware.

Originally mentioned here: https://github.com/laravel/framework/issues/8202

### CODE REVIEW

File: src/Illuminate/Foundation/Http/Middleware/VerifyPostSize.php [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+	 *
+	 * @param  \Illuminate\Http\Request  $request
+	 * @param  \Closure  $next
+	 * @return mixed
+	 * 
+	 * @throws \Illuminate\Http\Exception\PostTooLargeException
+	 */
+	public function handle($request, Closure $next)
+	{
+		if (isset($_SERVER['CONTENT_LENGTH']) && ($_SERVER['CONTENT_LENGTH'] > ini_get('post_max_size')))
```
Comment (franzliedke) [inline src/Illuminate/Foundation/Http/Middleware/VerifyPostSize.php]: You should retrieve these from `$request->server`, not from the superglobals...

Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Http/Middleware/VerifyPostSize.php]: Yeh, @franzliedke's absolutely correct.

**Comment by rweas** [comment]
I realize that using laravel's classes like Request is important instead of super globals but why is that exactly?

**Comment by GrahamCampbell** [comment]
> I realize that using laravel's classes like Request is important instead of super globals but why is that exactly?

You say you realise that it's important, then ask why - that's not realising it's important?

The reason is that you MUST NOT be making assumptions about the global state of php. Encapsulating all information about a request allows you to serve multiple requests from a single php instance, for example, during testing.

**Comment by TheShiftExchange** [comment]
Ok - I've made some changes. There was actually a bug - so this fixes that.

The `getPostMaxSize()` function is drawn from the official PHP docs on how to get the value as bytes: http://php.net/manual/en/function.ini-get.php

File: src/Illuminate/Foundation/Http/Middleware/VerifyPostSize.php [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+	}
+	
+	/**
+	 * Determine the server 'post_max_size' as bytes
+	 *
+	 * @return int
+	 */
+	protected function getPostMaxSize()
+	{
+		$post_max_size = trim(ini_get('post_max_size'));
```
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Http/Middleware/VerifyPostSize.php]: You must use camelCase, not snake_case.

**Comment by TheShiftExchange** [comment]
Reading further down the php.net page - there is an optimised function. Given we'll be calling this on every call - thought might as well switch to that.

I'll squash all this to one commit once everyone is happy (and assuming Taylor even wants to accept ths).

File: src/Illuminate/Foundation/Http/Middleware/VerifyPostSize.php [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+	/**
+	 * Determine the server 'post_max_size' as bytes
+	 *
+	 * @return int
+	 */
+   	protected function getPostMaxSize()
+   	{
+		$postMaxSize = ini_get('post_max_size');
+		
+		switch (substr ($postMaxSize, -1)) {
```
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Http/Middleware/VerifyPostSize.php]: Please fix all the cs issues. Every line of this method is wrong from here down.

File: src/Illuminate/Foundation/Http/Middleware/VerifyPostSize.php [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+		if ($request->server('CONTENT_LENGTH') > $this->getPostMaxSize())
+		{
+			throw new PostTooLargeException;
+		}
+
+		return $next($request);
+	}
+	
+	/**
+	 * Determine the server 'post_max_size' as bytes
```
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Http/Middleware/VerifyPostSize.php]: Missing fullstop.

**Comment by TheShiftExchange** [comment]
ping @GrahamCampbell - is that better?

File: src/Illuminate/Foundation/Http/Middleware/VerifyPostSize.php [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+			case 'K':
+			case 'k':
+				return (int)$postMaxSize * 1024;
+				
+			case 'G':
+			case 'g':
+				return (int)$postMaxSize * 1073741824;
+		}
+		
+		return (int)$postMaxSize;
```
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Http/Middleware/VerifyPostSize.php]: Missing space.

File: src/Illuminate/Foundation/Http/Middleware/VerifyPostSize.php [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+	}
+}
```
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Http/Middleware/VerifyPostSize.php]: You need an empty line before the closing class brace.

**Comment by GrahamCampbell** [comment]
Fix the two remaining things and squash to one commit please.

File: src/Illuminate/Foundation/Http/Middleware/VerifyPostSize.php (hunk 5 above)
Comment (TheShiftExchange) [inline src/Illuminate/Foundation/Http/Middleware/VerifyPostSize.php]: Sorry - which space where? I dont know what you mean...

Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Http/Middleware/VerifyPostSize.php]: After `(int)`, you need a space.

Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Http/Middleware/VerifyPostSize.php]: Same for the above bits too.

File: src/Illuminate/Foundation/Http/Middleware/VerifyPostSize.php [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+   	protected function getPostMaxSize()
+   	{
+		$postMaxSize = ini_get('post_max_size');
+		
+		switch (substr($postMaxSize, -1))
+		{
+			case 'M':
+			case 'm':
+				return (int)$postMaxSize * 1048576;
+
```
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Http/Middleware/VerifyPostSize.php]: don't leave blank lines

File: src/Illuminate/Foundation/Http/Middleware/VerifyPostSize.php [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+		switch (substr($postMaxSize, -1))
+		{
+			case 'M':
+			case 'm':
+				return (int)$postMaxSize * 1048576;
+				
+			case 'K':
+			case 'k':
+				return (int)$postMaxSize * 1024;
+
```
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Http/Middleware/VerifyPostSize.php]: don't leave blank lines

File: src/Illuminate/Foundation/Http/Middleware/VerifyPostSize.php [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+				
+			case 'K':
+			case 'k':
+				return (int)$postMaxSize * 1024;
+				
+			case 'G':
+			case 'g':
+				return (int)$postMaxSize * 1073741824;
+		}
+
```
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Http/Middleware/VerifyPostSize.php]: this blank line is good, leave it

File: src/Illuminate/Foundation/Http/Middleware/VerifyPostSize.php (hunk 7 above)
Comment (TheShiftExchange) [inline src/Illuminate/Foundation/Http/Middleware/VerifyPostSize.php]: All the current Illuminate `switch` code has blank lines. Thats where I copied this style code from.

**Comment by TheShiftExchange** [comment]
Sigh... I'll fix this - I obviously did something wrong squashing to one commit... working on it now...

**Comment by GrahamCampbell** [comment]
You rebased against the master instead of 5.1 by the looks of it.

**Comment by GrahamCampbell** [comment]
You rebased against the master instead of 5.0 by the looks of it.

**Comment by TheShiftExchange** [comment]
Yeah - my bad. Still learning Git properly. Anyway - for simplicity I've done another PR - https://github.com/laravel/framework/pull/8215

**Pull request closed without merging**